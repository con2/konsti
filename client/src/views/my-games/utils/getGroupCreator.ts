import { GroupMember } from "shared/typings/api/groups";

export const getGroupCreator = (
  groupMembers: readonly GroupMember[]
): GroupMember | null => {
  const groupCreator = groupMembers.find(
    (member) => member.serial === member.groupCode
  );
  if (!groupCreator) return null;
  return groupCreator;
};
