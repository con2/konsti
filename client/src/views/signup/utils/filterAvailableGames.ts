import _ from 'lodash';
import { config } from 'config';
import { Game } from 'typings/game.typings';
import { Signup } from 'typings/user.typings';

// Get games that have are not hidden, have signup open, and are not signed
export const filterAvailableGames = (
  games: readonly Game[],
  hiddenGames: readonly Game[],
  selectedGames: readonly Signup[],
  signupTime: string
): readonly Game[] => {
  const visibleGames = _.differenceBy(games, hiddenGames, 'gameId');

  const noSignupGames = config.noSignupGames;

  const signupOpenGames = visibleGames.filter((game) => {
    if (noSignupGames.includes(game.gameId)) return false;
    else return true;
  });

  const selectedGameDetails = selectedGames.map(
    (selectedGame) => selectedGame.gameDetails
  );

  const availableGames = _.differenceBy(
    signupOpenGames,
    selectedGameDetails,
    'gameId'
  );

  const availableGamesForStartTime = availableGames.filter(
    (nonSelectedGame) => nonSelectedGame.startTime === signupTime
  );

  return _.sortBy(availableGamesForStartTime, [
    (game) => game.title.toLowerCase(),
  ]);
};
