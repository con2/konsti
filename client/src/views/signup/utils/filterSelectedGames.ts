import { Signup } from 'typings/user.typings';
import { Game } from 'typings/game.typings';

export const filterSelectedGames = (
  selectedGames: readonly Signup[],
  signupTime: string
): readonly Game[] => {
  const selectedGameDetails = selectedGames.map(
    (selectedGame) => selectedGame.gameDetails
  );

  return selectedGameDetails.filter(
    (selectedGame) => selectedGame.startTime === signupTime
  );
};
