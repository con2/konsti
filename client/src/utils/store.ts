import {
  createStore,
  combineReducers,
  applyMiddleware,
  CombinedState,
  AnyAction,
} from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { config } from 'config';
import { loadSession } from 'utils/localStorage';
import { RootState } from 'typings/redux.typings';
import { SUBMIT_LOGOUT } from 'typings/logoutActions.typings';

// Reducers
import { reducer as formReducer } from 'redux-form';
import { allGamesReducer } from 'views/all-games/allGamesReducer';
import { loginReducer } from 'views/login/loginReducer';
import { myGamesReducer } from 'views/my-games/myGamesReducer';
import { signupReducer } from 'views/signup/signupReducer';
import { adminReducer } from 'views/admin/adminReducer';
import { resultsReducer } from 'views/results/resultsReducer';

// Set reducers
export const appReducer = combineReducers({
  form: formReducer,
  allGames: allGamesReducer,
  login: loginReducer,
  myGames: myGamesReducer,
  signup: signupReducer,
  admin: adminReducer,
  results: resultsReducer,
});

// Reducer to reset state
const rootReducer = (
  state: RootState,
  action: AnyAction
): CombinedState<RootState> => {
  if (action.type !== SUBMIT_LOGOUT) return appReducer(state, action);

  const newState = appReducer(undefined, action);
  newState.admin = state.admin;
  newState.allGames = state.allGames;
  return newState;
};

const middlewares = applyMiddleware(thunk);

const composeEnhancers = composeWithDevTools({
  trace: config.reduxTrace,
  traceLimit: 25,
});

const enhancer = composeEnhancers(middlewares);

// Load persisted state from localStorage
const persistedState = loadSession();

// Create a Redux store object that holds the app state
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
export const store = createStore(rootReducer, persistedState, enhancer);
