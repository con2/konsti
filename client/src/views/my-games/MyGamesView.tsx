import React, { FC, ReactElement } from 'react';
import { useSelector, useStore } from 'react-redux';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { MySignupsList } from 'views/my-games/components/MySignupsList';
import { MyFavoritesList } from 'views/my-games/components/MyFavoritesList';
import { MyEnteredList } from 'views/my-games/components/MyEnteredList';
import {
  getUpcomingSignedGames,
  getUpcomingEnteredGames,
  getUpcomingFavorites,
} from 'utils/getUpcomingGames';
import { loadUser, loadGames, loadGroupMembers } from 'utils/loadData';
import { isGroupLeader } from 'views/group/GroupView';
import { Game } from 'typings/game.typings';
import { Signup } from 'typings/user.typings';
import { GroupMember } from 'typings/group.typings';
import { RootState } from 'typings/redux.typings';

export const MyGamesView: FC = (): ReactElement => {
  const { t } = useTranslation();

  const serial: string = useSelector((state: RootState) => state.login.serial);
  const groupCode: string = useSelector(
    (state: RootState) => state.login.groupCode
  );
  const signedGames: readonly Signup[] = useSelector(
    (state: RootState) => state.myGames.signedGames
  );
  const favoritedGames: readonly Game[] = useSelector(
    (state: RootState) => state.myGames.favoritedGames
  );
  const enteredGames: readonly Signup[] = useSelector(
    (state: RootState) => state.myGames.enteredGames
  );
  const groupMembers: readonly GroupMember[] = useSelector(
    (state: RootState) => state.login.groupMembers
  );
  const testTime: string = useSelector(
    (state: RootState) => state.admin.testTime
  );

  const [showAllGames, setShowAllGames] = React.useState<boolean>(false);

  const store = useStore();

  React.useEffect(() => {
    const fetchData = async (): Promise<void> => {
      await loadGames();
      await loadUser();
      await loadGroupMembers();
    };
    fetchData();
  }, [store, testTime]);

  const getGroupLeader = (
    groupMembers: readonly GroupMember[]
  ): GroupMember | null => {
    const groupLeader = groupMembers.find(
      (member) => member.serial === member.groupCode
    );
    if (!groupLeader) return null;
    return groupLeader;
  };

  const getSignedGames = (
    signedGames: readonly Signup[]
  ): readonly Signup[] => {
    if (isGroupLeader(groupCode, serial)) {
      if (!showAllGames) return getUpcomingSignedGames(signedGames);
      else return signedGames;
    }

    if (!isGroupLeader(groupCode, serial)) {
      const groupLeader = getGroupLeader(groupMembers);

      if (!showAllGames) {
        return getUpcomingSignedGames(
          groupLeader ? groupLeader.signedGames : signedGames
        );
      } else return groupLeader ? groupLeader.signedGames : signedGames;
    }

    return signedGames;
  };

  return (
    <div className='my-games-view'>
      <>
        <div className='my-games-toggle-visibility'>
          <button
            onClick={() => setShowAllGames(false)}
            disabled={!showAllGames}
          >
            {t('lastStartedAndUpcomingGames')}
          </button>
          <button onClick={() => setShowAllGames(true)} disabled={showAllGames}>
            {t('allGames')}
          </button>
        </div>

        <MyFavoritesList
          favoritedGames={
            showAllGames ? favoritedGames : getUpcomingFavorites(favoritedGames)
          }
        />

        {!isGroupLeader(groupCode, serial) && (
          <MyGamesGroupNotification>
            <p className='bold'>{t('inGroupSignups')}</p>
          </MyGamesGroupNotification>
        )}

        <MySignupsList signedGames={getSignedGames(signedGames)} />

        <MyEnteredList
          enteredGames={
            showAllGames ? enteredGames : getUpcomingEnteredGames(enteredGames)
          }
          signedGames={getSignedGames(signedGames)}
        />
      </>
    </div>
  );
};

const MyGamesGroupNotification = styled.div`
  margin: 30px 0 0 0;
`;
