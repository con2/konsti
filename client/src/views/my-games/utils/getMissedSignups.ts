import moment from 'moment';
import { getStartTimes } from 'utils/getStartTimes';
import { config } from 'config';
import { Signup } from 'typings/user.typings';
import { getTime } from 'utils/getTime';

export const getMissedSignups = (
  signedGames: readonly Signup[],
  enteredGames: readonly Signup[]
): string[] => {
  const timeNow = getTime();

  const signedGamesStartTimes = getStartTimes(
    signedGames.map((signedGame) => signedGame.gameDetails)
  );

  // Get signup times for past signed games
  const pastSignupTimes = signedGamesStartTimes.filter(
    (signedGamesStartTime) => {
      const signupEndTime = moment(signedGamesStartTime).subtract(
        config.SIGNUP_END_TIME,
        'minutes'
      );

      if (signupEndTime.isBefore(moment(timeNow))) {
        return signedGamesStartTime;
      }
    }
  );

  // Check if there are past signed games without entered game => missed signup
  const missedSignupTimes = pastSignupTimes.filter((pastSignupTime) => {
    let found = false;
    if (enteredGames.length === 0) {
      return pastSignupTime;
    }

    enteredGames.find((enteredGame) => {
      if (enteredGame.time === pastSignupTime) {
        found = true;
      }
    });

    if (!found) {
      return pastSignupTime;
    }
  });

  const missedSignups: string[] = missedSignupTimes.map((missedSignupTime) => {
    return missedSignupTime;
  });

  return missedSignups;
};
