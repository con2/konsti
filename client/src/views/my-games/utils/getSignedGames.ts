import { getUpcomingSignedGames } from "client/utils/getUpcomingGames";
import { getIsGroupCreator } from "client/views/group/groupUtils";
import { getGroupCreator } from "client/views/my-games/utils/getGroupCreator";
import { GroupMember } from "shared/typings/api/groups";
import { SelectedGame } from "shared/typings/models/user";

export const getSignedGames = (
  signedGames: readonly SelectedGame[],
  groupCode: string,
  serial: string,
  showAllGames: boolean,
  groupMembers: readonly GroupMember[]
): readonly SelectedGame[] => {
  const isGroupCreator = getIsGroupCreator(groupCode, serial);

  if (isGroupCreator) {
    if (!showAllGames) return getUpcomingSignedGames(signedGames);
    else return signedGames;
  }

  if (!isGroupCreator) {
    const groupCreator = getGroupCreator(groupMembers);

    if (!showAllGames) {
      return getUpcomingSignedGames(
        groupCreator ? groupCreator.signedGames : signedGames
      );
    } else return groupCreator ? groupCreator.signedGames : signedGames;
  }

  return signedGames;
};
