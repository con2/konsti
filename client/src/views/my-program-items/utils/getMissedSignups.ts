import dayjs from "dayjs";
import { getStartTimes } from "client/utils/getStartTimes";
import { config } from "shared/config";
import { getTimeNow } from "client/utils/getTimeNow";
import { Signup } from "shared/types/models/user";

export const getMissedSignups = (
  lotterySignups: readonly Signup[],
  directSignups: readonly Signup[],
): string[] => {
  // Wait this long before showing "you didn't get into program item"
  // TODO: Instead of hard-coding, figure dynamically if assignment is still running
  const safePeriod = 3; // minutes

  const timeNow = getTimeNow();

  const lotterySignupsStartTimes = getStartTimes(
    lotterySignups.map((lotterySignup) => lotterySignup.programItemDetails),
  );

  // Get signup times for past lottery signup
  const pastSignupTimes = lotterySignupsStartTimes.filter(
    (lotterySignupsStartTime) => {
      const signupEndTime = dayjs(lotterySignupsStartTime).subtract(
        config.shared().DIRECT_SIGNUP_START - safePeriod,
        "minutes",
      );

      if (signupEndTime.isBefore(dayjs(timeNow))) {
        return lotterySignupsStartTime;
      }
    },
  );

  // Check if there are past lottery signups without direct signup => missed signup
  const missedSignupTimes = pastSignupTimes.filter((pastSignupTime) => {
    let found = false;
    if (directSignups.length === 0) {
      return pastSignupTime;
    }

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    directSignups.find((directSignup) => {
      if (directSignup.time === pastSignupTime) {
        found = true;
      }
    });

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!found) {
      return pastSignupTime;
    }
  });

  const missedSignups: string[] = missedSignupTimes.map((missedSignupTime) => {
    return missedSignupTime;
  });

  return missedSignups;
};
