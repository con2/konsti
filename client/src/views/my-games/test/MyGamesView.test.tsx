import React from 'react';
import { Provider } from 'react-redux';
import { store } from 'utils/store';
import { shallow } from 'enzyme';
import { MyGamesView } from '../MyGamesView';

describe('MyGamesView', () => {
  it('should render correctly', () => {
    const component = shallow(
      <Provider store={store}>
        <MyGamesView />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
