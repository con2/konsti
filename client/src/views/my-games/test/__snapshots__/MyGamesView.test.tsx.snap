// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyGamesView should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MyGamesView />
</ContextProvider>
`;
